### PR TITLE
internal/v5: allow upload of docker images as resources

### PR DESCRIPTION
**Pull request by rogpeppe**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2206/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2207/

File: internal/v5/resources.go [hunk 1]
```diff
@@ -40,6 +46,42 @@ func (h *ReqHandler) serveResources(id *router.ResolvedURL, w http.ResponseWrite
...
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	if !charmstore.IsKubernetesCharm(e.CharmMeta) {
+		return errgo.WithCausef(nil, params.ErrForbidden, "%q does not support docker resource upload", id.URL.String())
+	}
+	r, ok := e.CharmMeta.Resources[resourceName]
+	if !ok {
+		return errgo.WithCausef(nil, params.ErrForbidden, "%q has no resource named %q", id.URL.String(), resourceName)
+	}
+	if r.Type != resource.TypeDocker {
```
Comment (fabricematrat) [inline internal/v5/resources.go:70]: Just checking but you can upload a k8s charm without specifying first it's resources ?

**Review by fabricematrat — APPROVED** [review]

File: config/config.go [hunk 2]
```diff
@@ -56,6 +56,7 @@ type Config struct {
 	SwiftTenant                    string            `yaml:"swift-tenant"`
 	SwiftAuthMode                  *SwiftAuthMode    `yaml:"swift-authmode"`
 	LoggingConfig                  string            `yaml:"logging-config"`
+	DockerRegistryAddress          string            `yaml:"docker-registry-address"`
```
Comment (mhilton) [inline config/config.go:59]: We should really add a test for this.

File: internal/charmstore/resources.go [hunk 3]
```diff
@@ -205,6 +206,31 @@ func (s *Store) AddResourceWithUploadId(id *router.ResolvedURL, name string, upl
 	return res, nil
 }
 
+func (s *Store) AddDockerResource(id *router.ResolvedURL, resourceName, imageName, digest string) (*mongodoc.Resource, error) {
```
Comment (mhilton) [inline internal/charmstore/resources.go]: Please add a doc comment

File: internal/charmstore/resources.go [hunk 4]
```diff
@@ -205,6 +206,31 @@ func (s *Store) AddResourceWithUploadId(id *router.ResolvedURL, name string, upl
...
+	entity, err := s.FindEntity(id, FieldSelector("charmmeta", "baseurl"))
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	if !IsKubernetesCharm(entity.CharmMeta) {
+		return nil, errgo.Newf("entity is not a kubernetes charm")
+	}
+	if !charmHasDockerResource(entity.CharmMeta, resourceName) {
+		return nil, errgo.Newf("%q does not have image resource %q", id, resourceName)
```
Comment (mhilton) [inline internal/charmstore/resources.go:221]: "does not have docker resource"?

File: internal/charmstore/server.go [hunk 5]
```diff
@@ -243,6 +247,7 @@ type Server struct {
 
 // ServeHTTP implements http.Handler.ServeHTTP.
 func (s *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	logger.Infof("Server.ServeHTTP %q", req.URL)
```
Comment (mhilton) [inline internal/charmstore/server.go]: This looks like a debugging remnant.

File: internal/charmstore/server.go [hunk 6]
```diff
@@ -271,5 +276,6 @@ func handle(mux *router.ServeMux, path string, handler http.Handler) {
 		handler = http.StripPrefix(path, handler)
 		path += "/"
 	}
+	logger.Infof("handling path %q with %T", path, handler)
```
Comment (mhilton) [inline internal/charmstore/server.go]: Another debugging remnant?

File: internal/charmstore/store.go [hunk 7]
```diff
@@ -994,6 +994,7 @@ func (s *Store) checkPublishedResources(entity *mongodoc.Entity, resources map[s
 	if err != nil {
 		return errgo.Mask(err)
 	}
+	logger.Infof("checkPublishedResources got resources %#v", knownResources)
```
Comment (mhilton) [inline internal/charmstore/store.go]: and again?

File: internal/dockerauth/api.go [hunk 8]
```diff
@@ -58,40 +53,17 @@ func parseScope(s string) ([]resourceAccessRights, error) {
...
 		return nil, nil
 	}
-	var errs []error
 	var ras []resourceAccessRights
 	for _, s := range strings.Split(s, " ") {
 		ra, err := parseResourceAccessRights(s)
 		if err != nil {
-			errs = append(errs, err)
-			continue
+			return nil, errgo.Notef(err, "invalid access rights %q", s)
```
Comment (mhilton) [inline internal/dockerauth/api.go]: "invalid scope" makes more sense as that's the parameter name.

File: internal/v5/resources_test.go [hunk 9]
```diff
@@ -402,6 +403,235 @@ func (s *ResourceSuite) TestUploadResourcePathWithNoExtension(c *gc.C) {
...
+		ExpectStatus: http.StatusBadRequest,
+		ExpectBody: params.Error{
+			Code:    params.ErrBadRequest,
+			Message: `must specify resource-name parameter`,
+		},
+		Do: s.bakeryDoAsUser("charmers"),
+	})
+}
+
+
```
Comment (mhilton) [inline internal/v5/resources_test.go:577]: you have a spare blank line here

**Review by mhilton — APPROVED** [review]
LGTM modulo a few thoughts

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2208/

File: config/config.go (hunk 2 above)
Comment (rogpeppe) [reply config/config.go:59]: done

File: internal/v5/resources_test.go (hunk 9 above)
Comment (rogpeppe) [reply internal/v5/resources_test.go:577]: done

File: internal/dockerauth/api.go (hunk 8 above)
Comment (rogpeppe) [reply internal/dockerauth/api.go]: done

File: internal/v5/resources.go (hunk 1 above)
Comment (rogpeppe) [reply internal/v5/resources.go:70]: Yeah; that's true of any charm - you can't upload resources to a charm that doesn't yet exist.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2209/

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.jujugui.org:8080/job/charmstore-merge/629

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2210/

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**